crates/renderer/src/tree_renderer.rs

```diff
@@ -558,63 +556,66 @@ impl TreeRenderer {
-                        tracing::trace!(?count, ?offset, "draw node primitives");
-                        for i in 0..count {
-                            render_pass.draw_indexed_indirect(
-                                collect_state.commands.buffer(),
-                                (offset + i as u64)
-                                    * std::mem::size_of::<DrawIndexedIndirect>() as u64,
-                            );
-                        }
+                if self.config.render_mode == RenderMode::MultiIndirect {
+                    #[cfg(not(any(target_os = "windows", target_os= "linux")))]
+                    {
+                        panic!("MultiIndirect is not supported on the current platform");
+                    }
+                    #[cfg(any(target_os = "windows", target_os= "linux"))]
+                    {
+                        render_pass.multi_draw_indexed_indirect_count(
+                            collect_state.commands.buffer(),
+                            offset * std::mem::size_of::<DrawIndexedIndirect>() as u64,
+                            collect_state.counts.buffer(),
+                            mat.material_index as u64 * std::mem::size_of::<u32>() as u64,
+                            mat.primitives.len() as u32,
+                        );
+                    }
+                } else if self.config.render_mode == RenderMode::Indirect {
+                    // If none, a new material has been added, but the async buffer read has
+                    // not yet finished and we are still using the previous frame's material
+                    // counts.
+                    let count = count_state
+                        .counts()
+                        .get(mat.material_index as usize)
+                        .copied()
+                        .unwrap_or(0);
+
+                    // NOTE: this issues 1 draw call *for every single visible primitive* in the scene
+                    tracing::trace!(?count, ?offset, "draw node primitives");
+                    for i in 0..count {
+                        render_pass.draw_indexed_indirect(
+                            collect_state.commands.buffer(),
+                            (offset + i as u64) * std::mem::size_of::<DrawIndexedIndirect>() as u64,
+                        );
+                    }
+                } else if self.config.render_mode == RenderMode::Direct {
+                    for (i, &(id, primitive_idx)) in mat.primitives.iter().enumerate() {
+                        let primitive = &world.get_ref(id, primitives()).unwrap()[primitive_idx];
+
+                        let mesh = mesh_buffer.get_mesh_metadata(&primitive.mesh);
+                        let index = offset + i as u64;
+
+                        render_pass.draw_indexed(
+                            mesh.index_offset..(mesh.index_offset + mesh.index_count),
+                            // mesh.base_offset as i32,
```

Comment:
Strange.

Adding `mesh.base_offset` makes the direct render mode work on the web, but stops it working on native (gray screen).

Doing 0 yields the other way around

🤔

Reply:
Update:

This line fixed direct rendering on windows web, but breaks direct rendering on mac web, as well as native